### ISSUE TITLE
--manual should display a warning about the inability to use certbot renew

### ISSUE THREAD

**Issue by schoen** [certbot/certbot#6280]
Recently, one of the most common areas of confusion on the forum has been when people who obtained certificates with `--manual` try to renew them with `certbot renew` and have a great deal of trouble understanding the text

```
The error was: PluginError(‘An authentication script must be provided with --manual-auth-hook when using the manual plugin non-interactively.’,)
Attempting to renew cert (example.com) from /etc/letsencrypt/renewal/example.com.conf produced an unexpected error: The manual plugin is not working; there may be problems with your existing configuration.
The error was: PluginError(‘An authentication script must be provided with --manual-auth-hook when using the manual plugin non-interactively.’,). Skipping.
```

It looks to me like one problem is that the only relevant information here is

`An authentication script must be provided with --manual-auth-hook when using the manual plugin non-interactively.`

which appears in the middle of a lot of other error output (and perhaps many users also don't understand that `certbot renew` uses authenticator plugins non-interactively).

We could try to improve this message a lot, but I would also like to see `--manual` show a warning at the beginning and end of the process when no auth-hook is provided: *This certificate cannot be renewed by `certbot renew`. This certificate requires manual human intervention for each renewal!* (or something similar). Maybe that will help some people realize that `--manual` wasn't really what they wanted, or at least give them some more useful context for the `certbot renew` errors.

In general, I think we should do more to warn people away from using `--manual` because it's not that likely to give them an experience that they'll like at renewal time (even apart from the specific error messages). Its popularity partly dates to a time when it was the mainstream way of using DNS-01, which in principle is no longer the case.

**Comment by gridworkz** [comment]
So this begs the bigger question - What certbot command syntax should I be using to generate certs (without the --manual switch) that will subsequently allow me to use certbot renew? Any examples? Cheers, Dave

**Comment by schoen** [comment]
Well, I wasn't necessarily expecting people to show up and ask about the underlying substantive question in this GitHub issue, but in answer to @askpatrickw and @gridworkz:

Most people use `--manual` because they need a wildcard certificate. If you don't need a wildcard certificate, you should most likely use a different authenticator, like `--apache`, `--nginx`, `--webroot`, or `--standalone`, depending on your needs. The other authenticators are much easier to use and support automated renewal out of the box (except for `--standalone` on systems where there's also an existing web server listening on port 80).

If you *do* need a wildcard certificate, the trouble is that you need to update your DNS records by posting new `_acme-challege` TXT records *every time the certificate is renewed*. Certbot isn't capable of doing this itself unless you give it some kind of code and credentials that will allow it to make changes to your DNS records; there's no general-purpose way of doing this that would work in every environment.

What you normally need in this case is a DNS provider API, and then one of the Certbot DNS authenticator plugins, which are not necessarily easy to install, depending on how you installed Certbot.

The existing DNS plugins are described and listed at https://certbot.eff.org/docs/using.html#dns-plugins, and there may be information about installing them in conjunction with your copy of Certbot at https://certbot.eff.org/, if you select information about your OS type.

If you're familiar with writing shell scripts, you can also write a "manual authenticator hook" script, which is a script that handles the authentication step for Certbot, for example by doing whatever is necessary to create the DNS records that Certbot requests. This possibility is described at https://certbot.eff.org/docs/using.html#hooks, which also lists the environment variables that Certbot will pass to the indicated script in order to tell it what it needs to do. *If* you obtain a certificate with `--manual` and provide the `--manual-auth-hook` option and an appropriate script, `certbot renew` will be able to renew that certificate. Otherwise, `certbot renew` can't renew certificates obtained with `--manual` because the process always requires human intervention, and `certbot renew` has no support for tasks that require human intervention and is designed to be run only non-interactively.

If you want to renew a certificate that you obtained with `--manual` with no associated `--manual-auth-hook` option, you should re-run your original Certbot command (e.g. `certbot certonly --manual -d "*.example.com" -d example.com`) when there are 30 days or fewer of validity remaining on the existing certificate, and repeat the process that you originally used to obtain it.

This, or a higher-level summary of it, is the information that we hope to make people aware of when they attempt to use `--manual` for the first time.

An ACME client with broader and better-integrated DNS provider API support is acme.sh (https://github.com/Neilpang/acme.sh), but if you use acme.sh you would still need to have a DNS provider that offers an API, and tell acme.sh your credentials for that API.

**Comment by gridworkz** [comment]
You nailed it - I now fully understand how --manual works.

**Comment by schoen** [comment]
Great! Now we just have to find a way to help all of our users make good decisions about this without reading this GitHub thread. :slightly_smiling_face:

**Comment by gridworkz** [comment]
lol - it's one of those gotchas that seems to get lost in the mix somehow.
It's almost as though you need a two separate tracks 1) with wildcard
domain processing 2) one with individual domain processing. It sort of gets
all lumped together and the distinction disappears.

**Comment by mohrezaei** [comment]
What is the rationale for the DNS challenge being a one-off each time? Asked differently, why shouldn't I be able to use ``--standalone`` with a static DNS challenge?

**Comment by mohrezaei** [comment]
Answering my own question: the problem with a static DNS entry is that any caller can then do a renewal and get the certificates. Is that correct?
I can think of straight forward ways to engineer around that, but I suspect it would require far too much change at let's encrypt's server side. Here is one solution (which is already in use by other similar schemes, such as DKIM): create a TXT DNS record with public key. The renewing client then provides a signed message (with a nonce) to renew a certificate, proving that it's in possession of the private key.

**Comment by adferrand** [comment]
I would like to add another strong argument in favor of separating the DNS challenge renewal purpose in a `our-new-fancy-roxy-plugin-with-very-appropriate-documentation-for-that-purpose` new plugin (yes I am in love with this name now).

I discussed some time ago for a POC in #6607 to add a manual delay hook. This POC was asserting a big problem with `manual` if DNS challenges: authenticator hooks are executed sequentially on each FQDN to be included in the SANs of the new certificate. This becomes a major problem when the authenticator takes a significant time to execute (for instance because it needs to wait for correct propagation of the new DNS record, or for the DNS API to be available again for new changes). Indeed in this case, having for instance 5 FQDN to validated, with 20 min each, will lead to a total execution of one hour to generate/renew a certificate.

However changing the logic to run in parallel, and make more appropriate operations (like delay) in new hooks, would be a major breaking change for the `manual` plugin. All of this new logic could be nicely integrated without backward compatible consideration if it is in a new plugin, with theses capabilities designed on its root.

**Comment by das-g** [comment]
> If you want to renew a certificate that you obtained with `--manual` with no associated `--manual-auth-hook` option, you should re-run your original Certbot command (e.g. `certbot certonly --manual -d "*.example.com" -d example.com`) when there are 30 days or fewer of validity remaining on the existing certificate, and repeat the process that you originally used to obtain it.

Does that mean that the [renewal instructions](https://docs.gitlab.com/ee/user/project/pages/lets_encrypt_for_gitlab_pages.html#renewal) for GitLab pages are incompatible to [their instructions for initially getting a certificate](https://docs.gitlab.com/ee/user/project/pages/lets_encrypt_for_gitlab_pages.html#obtaining-a-lets-encrypt-certificate) and thus wrong?

Also, it seems like that's a use-case that hasn't been properly considered yet:
- I let my domain point to a static webhosting service (like GitLab pages),
  - where I can control the (static) content and can provide a certificate to be used but can't run arbitrary code (So I can't run certbot there) 
     and
  - that doesn't provide built-in let's-encrypt support
- I thus initiate the challenge from my local machine, to which my domain doesn't point (and which might not even be reachable from the public internet)
- I copy specific content onto the static webhosting service to complete the challenge
- I provide the static webhosting service with the certificate, so that it uses it for my content

Some way of attended (=interactive) but semi-automated renewal would be nice for that scenario, too, because I might not remember the exact commands I used to initially produce the certificate(s), but all necessary information would already be in the config, if I understood correctly.

**Comment by schoen** [comment]
@das-g,

> Does that mean that the renewal instructions for GitLab pages are incompatible to their instructions for initially getting a certificate and thus wrong?

Yes. the `certbot renew` command that they suggest will not work with certificates that were obtained with a bare `-a manual`. So their instructions are inconsistent as specified.

**Comment by bmw** [comment]
> We could try to improve this message a lot, but I would also like to see --manual show a warning at the beginning and end of the process when no auth-hook is provided: This certificate cannot be renewed by certbot renew.

Doing this at the end of the process is currently quite easy. I personally think doing it here is more important anyway to avoid the possibility of the initial output scrolling off the screen before the user had a chance to read it.

Here's an example of adding a message to be printed at the end of Certbot's execution from the Apache plugin: https://github.com/certbot/certbot/blob/7d61e9ea56fffea2b6c49bce4f4b21af1cef28ab/certbot-apache/certbot_apache/configurator.py#L1306

> In general, I think we should do more to warn people away from using --manual because it's not that likely to give them an experience that they'll like at renewal time (even apart from the specific error messages).

We should probably also add a big, scary warning to https://certbot.eff.org/docs/using.html#manual describing this problem.

Is there anything else people think we should do? The manual plugin is still currently hidden from Certbot's UI meaning that you can never select the manual plugin interactively. You have to request it on the command line.

**Comment by das-g** [comment]
> Is there anything else people think we should do?

Probably out of scope for this issue (#6280), but I think: An easy and obvious way should be offered to re-use the settings of the initial certificate request for subsequent renewals, even if the the initial certificate request was done with `--manual`. Use cases for that seem to be:
* wildcard cert
* cert for statically hosted site

**Comment by schoen** [comment]
Hi @das-g,

Unfortunately, this is guaranteed to fail because the certificate renewal will always provide a new challenge to be completed to re-confirm control over the domain name. But the Certbot `manual` plugin *doesn't contain any code to complete these challenges*.

https://github.com/certbot/certbot/blob/master/certbot/plugins/manual.py#L150

(The only exception is when the user has already provided a custom script.)

As a result, running this function non-interactively would never allow the plugin to complete the challenge.

**Comment by das-g** [comment]
That's why I wrote "renewal" and not `certbot renew`. Off course it can't be completely non-interactive in these cases, but it could either not ask about all the settings that are already known and just stop for the user to manually fulfill the new challenge, or it could at least offer the previous settings as suggestions to the interactive questions.

**Comment by schoen** [comment]
That's already the default behavior of `certbot certonly`. If you run `certbot certonly --cert-name example.com`, and `example.com` is an existing certificate that was obtained with `--manual`, then if it's within the renewal period, it will only prompt for the user to complete the challenge(s), and not prompt for anything else.

**Comment by mohrezaei** [comment]
This issue is still relevant. There is no static, non-http based method to obtain certificates via `--manual`. It might be better to split this issue into several others, such as:

- Provide a static DNS based renewal method, e.g. by putting a public key in a TXT record.
- Use existing certificate to get a new certificate, so `--manual` is only required for the initial certificate. 
- Add top level howtos and guided step-by-step instructions for non-HTTP use cases. There are dozens of protocols that use TLS.

**Comment by Adspectus** [comment]
Actually, I am wondering why it is not just possible to skip certificates with the `authenticator=manual` in its renewal file from the `certbot renew` command. Or, provide a switch/option to control this. It is very annoying to get the error 2 times a day when the `certbot renew` is performed and the manual cert is due within 30 days. Or is there a workaround which I have not found yet?

**Comment by alexzorin** [comment]
I'm going to close this because Certbot now warns, upon the issuance of the certificate and also in the User Guide, that `--manual` certificates (without a hook) are not automatically renewing. 

It might also be an idea to add a shortlink to this error message which links to https://eff-certbot.readthedocs.io/en/stable/using.html#manual, we'd accept a PR for that:

https://github.com/certbot/certbot/blob/ed6bbde38fc94bcce0dfd797187b1efe801b274d/certbot/certbot/_internal/plugins/manual.py#L116-L117
